
```
#统计fasta文件有多少条序列
grep ‘>’ S_pastorianus.fasta | wc –l
#统计fastq文件中reads数目
awk '{s++}END{print s/4}' file.fastq
# 对fasta文件进行重新编号
awk '{ if( $0 ~ />/){num++;print ">PBS.F1Y_Tag"num;}else{print $0}}' PBS.F1Y.FinalTag.fa.tmp > PBS.F1Y.FinalTag.fa
```
## fasta序列提取
```
# --- 根据ID提取FASTA文件中的seq
#利用QIIME软件中的extract_seqs_by_sample_id.py脚本
extract_seqs_by_sample_id.py -i polished_assembly_r.fasta -o Monascus_scf397.fasta -s scaffold397
# --- 根据ID提取fastq或者fasta文件中的序列
seqtk subseq in.fq name.lst > out.fq
```
seqtk的用法
```
Usage:   seqtk <command> <arguments>
Version: 1.2-r102-dirty
Command: seq       common transformation of FASTA/Q
         comp      get the nucleotide composition of FASTA/Q
         sample    subsample sequences
         subseq    extract subsequences from FASTA/Q
         fqchk     fastq QC (base/quality summary)
         mergepe   interleave two PE FASTA/Q files
         trimfq    trim FASTQ using the Phred algorithm
         hety      regional heterozygosity
         gc        identify high- or low-GC regions
         mutfa     point mutate FASTA at specified positions
         mergefa   merge two FASTA/Q files
         famask    apply a X-coded FASTA to a source FASTA
         dropse    drop unpaired from interleaved PE FASTA/Q
         rename    rename sequence names
         randbase  choose a random base from hets
         cutN      cut sequence at long N
         listhet   extract the position of each het
```
## 多行转一行
```
# 把所有的换行符替换为空格
# tr这个命令，前面提到过，若想不起来 `man tr`查看
# 主意第二个参数，引号内为空格。
sed 's/^\(>.*\)/\1\t/' test.fasta | tr '\n' ' '
>SOX2     ACGAGGGACGCATCGGACGACTGCAGGACTGTC ACGAGGGACGCATCGGACGACTGCAGGACTGTC ACGAGGGACGCATCGGACGACTGCAGGAC >POU5F1     CGGAAGGTAGTCGTCAGTGCAGCGAGTCCGT CGGAAGGTAGTCGTCAGTGCAGCGAGTCC >NANOG     ACGAGGGACGCATCGGACGACTGCAGGACTGTC ACGAGGGACGCATCGGACGACTGCAGG ACGAGGGACGCATCGGACGACTGCAGGACTGTC ACGAGGGACGCATCGGACGACTGCAGGACTGT 

# 把最后一个空格替换为换行符
sed 's/^\(>.*\)/\1\t/' test.fasta | tr '\n' ' ' | sed -e 's/ $/\n/'
>SOX2     ACGAGGGACGCATCGGACGACTGCAGGACTGTC ACGAGGGACGCATCGGACGACTGCAGGACTGTC ACGAGGGACGCATCGGACGACTGCAGGAC >POU5F1     CGGAAGGTAGTCGTCAGTGCAGCGAGTCCGT CGGAAGGTAGTCGTCAGTGCAGCGAGTCC >NANOG     ACGAGGGACGCATCGGACGACTGCAGGACTGTC ACGAGGGACGCATCGGACGACTGCAGG ACGAGGGACGCATCGGACGACTGCAGGACTGTC ACGAGGGACGCATCGGACGACTGCAGGACTGT

# 把  ' >'替换为换行符 注意被替换的是 空格+大于号
# 当连用多个替换命令时，使用-e 隔开
sed 's/^\(>.*\)/\1\t/' test.fasta | tr '\n' ' ' | sed -e 's/ $/\n/' -e 's/ >/\n>/g'
>SOX2     ACGAGGGACGCATCGGACGACTGCAGGACTGTC ACGAGGGACGCATCGGACGACTGCAGGACTGTC ACGAGGGACGCATCGGACGACTGCAGGAC
>POU5F1     CGGAAGGTAGTCGTCAGTGCAGCGAGTCCGT CGGAAGGTAGTCGTCAGTGCAGCGAGTCC
>NANOG     ACGAGGGACGCATCGGACGACTGCAGGACTGTC ACGAGGGACGCATCGGACGACTGCAGG ACGAGGGACGCATCGGACGACTGCAGGACTGTC ACGAGGGACGCATCGGACGACTGCAGGACTGT

# 把所有的空格替换掉
sed 's/^\(>.*\)/\1\t/' test.fasta | tr '\n' ' ' | sed -e 's/ $/\n/' -e 's/ >/\n>/g' -e 's/ //g'
>SOX2    ACGAGGGACGCATCGGACGACTGCAGGACTGTCACGAGGGACGCATCGGACGACTGCAGGACTGTCACGAGGGACGCATCGGACGACTGCAGGAC
>POU5F1    CGGAAGGTAGTCGTCAGTGCAGCGAGTCCGTCGGAAGGTAGTCGTCAGTGCAGCGAGTCC
>NANOG    ACGAGGGACGCATCGGACGACTGCAGGACTGTCACGAGGGACGCATCGGACGACTGCAGGACGAGGGACGCATCGGACGACTGCAGGACTGTCACGAGGGACGCATCGGACGACTGCAGGACTGT

# 把TAB键转换为换行符
sed 's/^\(>.*\)/\1\t/' test.fasta | tr '\n' ' ' | sed -e 's/ $/\n/' -e 's/ >/\n>/g' -e 's/ //g' -e 's/\t/\n/g' 
>SOX2
ACGAGGGACGCATCGGACGACTGCAGGACTGTCACGAGGGACGCATCGGACGACTGCAGGACTGTCACGAGGGACGCATCGGACGACTGCAGGAC
>POU5F1
CGGAAGGTAGTCGTCAGTGCAGCGAGTCCGTCGGAAGGTAGTCGTCAGTGCAGCGAGTCC

# 也可以直接用awk
# 差别只在一点
# 对于单行fasta文件，只需要记录一行，seq[name]=$0
# 对于多行fasta文件，需要把每一行序列都加到前面的序列上，seq[name]=seq[name]$0
awk 'BEGIN{OFS=FS="\t"}{if($0~/>/) {name=$0; sub(">", "", name);} else seq[name]=seq[name]$0;}END{print ">SOX2"; print seq["SOX2"]}' test.fasta
>SOX2
ACGAGGGACGCATCGGACGACTGCAGGACTGTCACGAGGGACGCATCGGACGACTGCAGGACTGTCACGAGGGACGCATCGGACGACTGCAGGAC
```

## GFF+fasta文件生成CDS序列:
```
awk '($3=="gene") {OFS="\t"; print $1,$4,$5}' spinosad_cluster.gff |  bedtools getfasta -fi spinosad_cluster.fa -bed - -fo spinosad_cluster.CDS.fa
```
## 统计fastq文件中总的碱基个数
```
# 统计fastq文件中总的碱基个数（包括N），有三种方法
le originA.rawdata_2.fq.gz |paste - - - - |cut -f 2|grep -o [ATCGNatcg]|wc -l
le originA.rawdata_2.fq.gz |awk 'BEGIN{sum=0;}{if(NR%4==2){sum+=length($0);}}END{print sum;}' -
le originA.rawdata_1.fq.gz| sed -n 2~4p | grep -o [ATCGNatcg]|wc -l
```
## 统计碱基数目和质量数的大小是否相等
```
# 处理fq文件，统计碱基数目和质量数的大小是否相等
for file in `cat data.list`
#for file in `cat data.list.head`
do
	less $file | paste - - - -| cut -f 2,4 | awk -F"\t" 'BEGIN {count=0;} {an[count] = $1;bn[count] = $2;count++;}; END{for (i = 0; i < NR; i++) if(length(an[i]) != length(bn[i])){ print filename, length(an[i]), length(bn[i])}}' filename="$file" >> result.list
done
```
## 计算fastq的质量值
```
less *.fastq.gz | head -n 1000 | awk '{if(NR%4==0) printf("%s",$0);}' | od -A n -t u1 -v | awk 'BEGIN{min=100;max=0;} {for(i=1;i<=NF;i++) {if($i>max) max=$i; if($i<min) min=$i;}}END \
{if(max<=126 && min<59) print "Phred33"; \
else if(max>73 && min>=64) print "Phred64"; \
else if(min>=59 && min<64 && max>73) print "Solexa64"; \
else print "Unknown score encoding"; }'
```
<!--stackedit_data:
eyJoaXN0b3J5IjpbMTg1NTM5NjUzNl19
-->